src/System.Linq.Parallel/tests/QueryOperators/ForAllTests.cs

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using Xunit;
+
+namespace Test
+{
+    public class ForAllTests
+    {
+        [Theory]
+        [MemberData("Ranges", (object)(new int[] { 1, 2, 16 }), MemberType = typeof(Sources))]
+        [MemberData("Ranges", (object)(new int[] { 1, 2, 16 }), MemberType = typeof(UnorderedSources))]
+        public static void ForAll(Labelled<ParallelQuery<int>> labelled, int count)
+        {
+            ParallelQuery<int> query = labelled.Item;
+            IntegerRangeSet seen = new IntegerRangeSet(0, count);
+            query.ForAll<int>(x => Assert.True(seen.Add(x)));
```

Comment:
> dies in the face of outerloop tests

I just grabbed your branch and tried it both as-is for inner loop and outer loop and then with a slight variant of your IntegerRangeSet:
```C#
internal class IntegerRangeSet : IEnumerable<KeyValuePair<int, bool>>
{
    private int[] _seen;
    private int _start;

    public IntegerRangeSet(int start, int count)
    {
        _start = start;
        _seen = new int[count];
    }

    public bool Add(int entry)
    {
        Assert.InRange(entry, _start, _start + _seen.Length - 1);
        return Interlocked.CompareExchange(ref _seen[entry - _start], 1, 0) == 0;
    }

    public bool Complete()
    {
        return _seen.All(x => x == 1);
    }
    ...
```
Prior to using Interlocked.CompareExchange, inner loop took 5 seconds and outer loop took 491 seconds.  After this switch, inner loop took 7 seconds and outer loop took 498.  That's only a 1.5% increase in outer loop time... you're seeing differently?  I could imagine using a lock with BitArray could be significantly more, but my comment about BitArray was just in passing, and before I realized you had concurrency problems.